File: frontend/src/api/contacts.ts
import { apiClient } from "./client";

export type ContactCreatePayload = {
  listing_id: string;
  full_name: string;
  phone_number: string;
  preferred_move_in_date?: string | null;
  message: string;
};

export type ContactCreateResponse = {
  id: string;
  listing_id: string;
  status: string;
  submitted_at: string;
};

export type ContactStatusResponse = {
  id: string;
  listing_id: string;
  status: "open" | "closed" | "confirmed";
  submitted_at: string;
  can_contact_again: boolean;
};

export async function createContact(payload: ContactCreatePayload): Promise<ContactCreateResponse> {
  const response = await apiClient.post<ContactCreateResponse>("/contacts", payload);
  return response.data;
}

export async function getContact(contactId: string): Promise<ContactStatusResponse> {
  const response = await apiClient.get<ContactStatusResponse>(`/contacts/${contactId}`);
  return response.data;
}

File: backend/app/schemas/contacts.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ContactCreateRequest(BaseModel):
    listing_id: str = Field(min_length=1)
    full_name: str = Field(min_length=2, max_length=80)
    phone_number: str = Field(min_length=7, max_length=32)
    preferred_move_in_date: date | None = None
    message: str = Field(min_length=10, max_length=500)

    @field_validator("listing_id", "full_name", "phone_number", "message")
    @classmethod
    def _trim_fields(cls, value: str) -> str:
        normalized = " ".join(value.strip().split())
        if not normalized:
            raise ValueError("Field cannot be empty")
        return normalized


class ContactCreateResponse(BaseModel):
    id: str
    listing_id: str
    status: str
    submitted_at: datetime


class ContactStatusResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    listing_id: str
    status: str
    submitted_at: datetime
    can_contact_again: bool
